aws/resource_aws_elasticsearch_domain.go

```diff
@@ -89,6 +89,24 @@ func resourceAwsElasticSearchDomain() *schema.Resource {
 					},
 				},
 			},
+			"encrypt_at_rest": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
```

Comment:
What purpose does this `bool` serve here? I think the behavior should be:
- If `kms_key_id` is given, create the ES with the `kms_key_id`
- If `kms_key_id` is not given, create the ES without the `kms_key_id`
- When things change, delete and create the es domain

Reply:
I was mirroring the API which takes those as options. I did consider setting it automatically when kms_key_id is set but it would prevent you from having Terraform encrypt it with the service KMS key that you get by default when using the console to create an encrypted domain.

The `aws_db_instance` resource has them as separate parameters to the resource (rather than an options block) and defaults to using the service KMS key if kms_key_id isn't specified but that's largely because the API looks that way.

Reply:
Yea true. Does AWS API create a default key if only `enable` is passed? If so, then having `enable` makes sense. If not, I am not sure whether kms key creation is within the scope of this resource.

For example, when adding triggers to lambda functions via aws console, aws does a lot of stuff in the background like creating IAM roles and adding policies and stuff. However in Terraform we don't do it and all the resources required to trigger a lambda function must be created.

Reply:
Yeah looks like it does, added a test that runs without specifying the key id and that works nicely without needing to do anything extra which is neat.

Reply:
Lovely! :)